### PR TITLE
feat: xud master password

### LINKED ISSUE #63: xud master password integration
Now that we pushed back the deadline, let's go for the integration of https://github.com/ExchangeUnion/xud/pull/1006. It'd be good to have some people test this on testnet.

After selecting testnet/mainnet - check if lndbtc wallet exists in `~/.xud-docker/testnet` / `~/.xud-docker/mainnet` (we'll make lndbtc a **must** in all later setups since lndbtc is required anyways to generate the seed).

If wallet does not exist:
* `No existing wallets found. Creating wallets...`
* Call `CreateNode`, `Please set a password (it secures xud and all your wallets):`
* *print potential errors* and exit
* If no errors, continue with `Please write down this 24 word mnemonic. You will be able to recover xud and funds of all your wallets with it, in case you forget your master password or loose your device. Keep it somewhere safe, it's your ONLY backup`: `CreateNode` output
* Did you write down your 24 word mnemonic? You will not be able to view it again after this prompt. [Y/n] n -> show again
* Y -> `All wallets successfully created.`
* start

If lndbtc wallet exists:
* call `unlock`, `Please enter your xud password:`
* start

### PR DESCRIPTION
**Pull request by reliveyy**
Resolves #63 

* Add xud master password support (xucli create)

### CODE REVIEW

**Comment by erkarl** [comment]
Why the removal of supervisor and moving tor to separate container? With this change we'll lose the auto restart process functionality.

**Comment by reliveyy** [comment]
> Why the removal of supervisor and moving tor to separate container? With this change we'll lose the auto restart process functionality.

https://github.com/ExchangeUnion/xud-docker/pull/134#issue-310302395 @erkarl

**Comment by erkarl** [comment]
LGTM in general. Needs manual testing. Could you please provide steps to test this:
`bash xud.sh -b feat/master-password` didn't seem to do the trick.

**Comment by kilrau** [comment]
`bash xud.sh -b feat/master-password` is working now, identified issues (not docker related):
1. Password is saved in cleartext in CLI history:
+ xucli='docker-compose exec xud xucli'
+ docker-compose exec xud xucli create 'MY_SECRET_PASSWORD'
2.  `create` returns empty mnemonic list:
+ echo 'Please write down this 24 word mnemonic. You will be able to recover funds of your wallets with it, in case you forget your master password or loose your device. Keep it somewhere safe, it is your ONLY backup'
Please write down this 24 word mnemonic. You will be able to recover funds of your wallets with it, in case you forget your master password or loose your device. Keep it somewhere safe, it is your ONLY backup

**Comment by kilrau** [comment]
Could you put the tor fix in a separate PR and enable it on testnet**+mainnet**? @reliveyy

File: main.sh [hunk 1]
```diff
@@ -66,14 +66,34 @@ log_details() {
...
+}
+
+check_wallets() {
+    if no_wallets; then
+        echo "Creating wallets..."
+        echo -n "Please set a password (it secures xud and all your wallets): "
+        read -s password
+        echo "Please write down this 24 word mnemonic. You will be able to recover funds of your wallets with it, in case you forget your master password or loose your device. Keep it somewhere safe, it is your ONLY backup"
+        xucli="docker-compose exec xud xucli"
+        if $xucli create "$password"; then
```
Comment (sangaman) [inline main.sh]: With the recent change to hide password input, we don't pass the password into the `create` call directly. So we also wouldn't need the prompt above to "please set a password" - although maybe the cli can be enhanced with a similar message in the parentheses.

File: main.sh [hunk 2]
```diff
@@ -66,14 +66,34 @@ log_details() {
...
+no_wallets() {
+    [[ ! -e data/lndbtc/data/chain/bitcoin/$network/wallet.db ]]
+}
+
+check_wallets() {
+    if no_wallets; then
+        echo "Creating wallets..."
+        echo -n "Please set a password (it secures xud and all your wallets): "
+        read -s password
+        echo "Please write down this 24 word mnemonic. You will be able to recover funds of your wallets with it, in case you forget your master password or loose your device. Keep it somewhere safe, it is your ONLY backup"
```
[the remark this answers is not in the document]
Comment (kilrau) [reply main.sh]: Agreed as just discussed in our call

**Comment by sangaman** [comment]
@reliveyy https://github.com/ExchangeUnion/xud/pull/1212 has been merged so you should be able to proceed with this off the latest master and remove the wording from this PR.

**Comment by kilrau** [comment]
The updating of the images was not the problem, after they are updated, I still have the same situation, see above.

**Comment by sangaman** [comment]
Looks like it needs to use an updated xud version that doesn't take the password from the command arguments.

**Comment by kilrau** [comment]
Looking at https://cloud.docker.com/u/exchangeunion/repository/registry-1.docker.io/exchangeunion/xud/tags, the image hasn't been updated in a while. Please let us know when all images are up-to-date and when to test again @reliveyy

**Comment by erkarl** [comment]
1. I'm getting the following output when testing this with locally built images:
```
Please choose the network: 2
Launching testnet environment...
D0919 19:22:58.252384704      31 env_linux.cc:71]            Warning: insecure environment read function 'getenv' used
You are creating an xud key and underlying wallets secured by a single password.
Enter a password:
Re-enter password:

xud was initialized without a seed because no wallets could be initialized.
All wallets successfully created.
```

2. Could we also rebase this branch against the current master?

File: main.sh [hunk 3]
```diff
@@ -64,14 +64,28 @@ log_details() {
...
+}
+
+check_wallets() {
+    if no_wallets; then
+        xucli="docker-compose exec xud xucli"
+        while ! $xucli getinfo | grep "UNIMPLEMENTED" >/dev/null 2>&1; do
+            sleep 3
+        done
+        if $xucli create; then
+            echo "All wallets successfully created."
```
Comment (kilrau) [inline main.sh]: This needs to check the actual response of xucli create. If `Passwords do not match, please try again`, then call xucli create again.

Comment (kilrau) [reply main.sh]: Also for `Error: 2 UNKNOWN: password must be at least 8 characters
All wallets successfully created.`

**Comment by reliveyy** [comment]
There are still some error handling issues:

- [ ] Password mistach
- [ ] Invalid password
- [ ] Ctrl-C (SIGINT)
- [ ] Re-invoke `xucli create`

**Comment by kilrau** [comment]
And another problem: even after waiting for several minutes before typing any password, it did not work. `status` shows that call clients are up and well:
```
kilrau@K-Yoga:~$ bash xud.sh -b feat/master-password
1) Simnet
2) Testnet
3) Mainnet
Please choose the network: 2
Launching testnet environment...
Pulling exchangeunion/lnd:0.7.1-beta__feat-master-password
Pulling exchangeunion/xud:latest__feat-master-password
Pulling exchangeunion/litecoind:0.17.1
Pulling exchangeunion/geth:1.9.1
Pulling exchangeunion/bitcoind:0.18.1
Pulling exchangeunion/raiden:latest
D0924 16:32:52.951447982      49 env_linux.cc:71]            Warning: insecure environment read function 'getenv' used
You are creating an xud key and underlying wallets secured by a single password.
Enter a password: 
Re-enter password: 

xud was initialized without a seed because no wallets could be initialized.
All wallets successfully created.

                           .___           __  .__   
          ___  _____ __  __| _/     _____/  |_|  |  
          \  \/  /  |  \/ __ |    _/ ___\   __\  |  
           >    <|  |  / /_/ |    \  \___|  | |  |__
          /__/\_ \____/\____ |     \___  >__| |____/
                \/          \/         \/           
--------------------------------------------------------------

testnet > docker-compose ps
       Name                      Command               State                                  Ports                               
----------------------------------------------------------------------------------------------------------------------------------
testnet_bitcoind_1    bitcoind -server -testnet  ...   Up      18332/tcp, 28332/tcp, 28333/tcp, 8332/tcp                          
testnet_geth_1        ./entrypoint.sh --syncmode ...   Up      30303/tcp, 30303/udp, 8545/tcp, 8546/tcp                           
testnet_litecoind_1   litecoind -server -testnet ...   Up      19332/tcp, 29332/tcp, 29333/tcp, 9332/tcp                          
testnet_lndbtc_1      /usr/bin/supervisord -c /e ...   Up      10009/tcp, 10735/tcp, 0.0.0.0:19735->19735/tcp, 20735/tcp, 9735/tcp
testnet_lndltc_1      /usr/bin/supervisord -c /e ...   Up      10009/tcp, 10735/tcp, 19735/tcp, 0.0.0.0:20735->20735/tcp, 9735/tcp
testnet_raiden_1      ./entrypoint.sh --keystore ...   Up      5001/tcp                                                           
testnet_xud_1         /bin/sh -c ./entrypoint.sh       Up      0.0.0.0:18885->18885/tcp, 8885/tcp, 8887/tcp                       
testnet > status
SERVICE	STATUS
btc	Syncing 23.09% (364810/1579541)
lndbtc	Waiting for sync
ltc	Syncing 0.86% (8096/938000)
lndltc	Waiting for sync
geth	Syncing 12.75% (823029/6451014)
raiden	Waiting for sync
xud	Waiting for sync
```

**Comment by kilrau** [comment]
NICE - It's finally working!

```
kilrau@K-Yoga:~$ xud -b feat/master-password
1) Simnet
2) Testnet
3) Mainnet
Please choose the network: 3
Launching mainnet environment...
Pulling exchangeunion/lnd:0.7.1-beta__feat-master-password
Pulling exchangeunion/xud:latest__feat-master-password
Pulling exchangeunion/litecoind:0.17.1
Pulling exchangeunion/geth:1.9.1
Pulling exchangeunion/bitcoind:0.18.1
Pulling exchangeunion/raiden:latest

.
 You are creating an xud key and underlying wallets secured by a single password.
Enter a password: 
Re-enter password: 

----------------------BEGIN XUD SEED---------------------
 1. abstract    2. jump        3. piece       4. only      
 5. regular     6. negative    7. logic       8. spider    
 9. over       10. artwork    11. still      12. glimpse   
13. depart     14. drift      15. owner      16. vacant    
17. wheel      18. skate      19. century    20. cancel    
21. crawl      22. pupil      23. push       24. craft     
-----------------------END XUD SEED----------------------

The following lnd wallets were initialized: BTC, LTC

Please write down your 24 word mnemonic. It will allow you to recover your xud
key and funds for the initialized wallets listed above should you forget your
password or lose your device. Keep it somewhere safe, it is your ONLY backup.
    


                           .___           __  .__   
          ___  _____ __  __| _/     _____/  |_|  |  
          \  \/  /  |  \/ __ |    _/ ___\   __\  |  
           >    <|  |  / /_/ |    \  \___|  | |  |__
          /__/\_ \____/\____ |     \___  >__| |____/
                \/          \/         \/           
--------------------------------------------------------------

mainnet >
```

**Comment by kilrau** [comment]
I see `xucli_create_wrapper.py` is outside the container though. Can it be placed within e.g. the xud container to avoid having a python dependency?

**Comment by kilrau** [comment]
Did run it like 10x. Worked 100% on Testnet, failed 100% on mainnet with `xud was initialized without a seed because no wallets could be initialized.` which doesn't make any sense to me. It's the exact same images and scripts.

**Comment by kilrau** [comment]
xud logs show grpc server shutdown in between createNode and node key creation:
```
xud_1        | 27/09/2019 08:57:48.885 [RPC] debug: received call /xudrpc.XudInit/CreateNode
xud_1        | 27/09/2019 08:57:48.909 [RPC] info: gRPC server completed shutdown
xud_1        | 27/09/2019 08:57:48.909 [GLOBAL] info: Local nodePubKey is 02fcd2da5fae54d34dd8b39c534ed84d0577f678ed6e7b6ba11f9b0a33a5ccf44c
xud_1        | 27/09/2019 08:57:48.924 [P2P] info: Connecting to mainnet XU network
xud_1        | 27/09/2019 08:57:48.926 [P2P] info: p2p server listening on 0.0.0.0:8885
xud_1        | 27/09/2019 08:57:48.927 [P2P] debug: Verifying reachability of advertised address: 0.0.0.0:8885
xud_1        | 27/09/2019 08:57:48.930 [HTTP] info: http server listening on 0.0.0.0:8887
xud_1        | 27/09/2019 08:57:48.933 [P2P] debug: Connected pre-handshake to 127.0.0.1:59734
xud_1        | 27/09/2019 08:57:48.991 [P2P] debug: Peer (02fcd2da5fae54d34dd8b39c534ed84d0577f678ed6e7b6ba11f9b0a33a5ccf44c): closing socket. reason: ConnectedToSelf
xud_1        | 27/09/2019 08:57:48.993 [P2P] warn: could not open connection to peer (02fcd2da5fae54d34dd8b39c534ed84d0577f678ed6e7b6ba11f9b0a33a5ccf44c): cannot attempt connection to self
xud_1        | 27/09/2019 08:57:48.994 [P2P] info: Peer 02fcd2da5fae54d34dd8b39c534ed84d0577f678ed6e7b6ba11f9b0a33a5ccf44c socket closed
xud_1        | 27/09/2019 08:57:48.997 [P2P] verbose: Verified reachability of advertised address: 0.0.0.0:8885
xud_1        | 27/09/2019 08:57:48.997 [RPC] info: gRPC server listening on 0.0.0.0:8886
```

**Comment by kilrau** [comment]
One more observation: I managed to have a case where only the ltc wallet initialization succeeded. Which does even make less sense to me since we use `LND-BTC` for seed creation.

```
xud_1        | 27/09/2019 09:22:24.999 [LND-LTC] info: wallet initialized
xud_1        | 27/09/2019 09:22:25.002 [LND-LTC] debug: watching /root/.lndltc/data/chain/litecoin/mainnet for admin.macaroon to be created
xud_1        | 27/09/2019 09:22:25.009 [RPC] info: gRPC server completed shutdown
xud_1        | 27/09/2019 09:22:25.009 [GLOBAL] info: Local nodePubKey is 0243f9ff226cda8d5f00a293c56f4406fc9d7ff80d2184bd986c8513756c6e9b61
xud_1        | 27/09/2019 09:22:25.041 [P2P] info: Connecting to mainnet XU network
xud_1        | 27/09/2019 09:22:25.046 [P2P] info: p2p server listening on 0.0.0.0:8885
xud_1        | 27/09/2019 09:22:25.048 [P2P] debug: Verifying reachability of advertised address: 0.0.0.0:8885
xud_1        | 27/09/2019 09:22:25.056 [HTTP] info: http server listening on 0.0.0.0:8887
xud_1        | 27/09/2019 09:22:25.061 [P2P] debug: Connected pre-handshake to 127.0.0.1:40472
xud_1        | 27/09/2019 09:22:25.168 [P2P] debug: Peer (0243f9ff226cda8d5f00a293c56f4406fc9d7ff80d2184bd986c8513756c6e9b61): closing socket. reason: ConnectedToSelf
xud_1        | 27/09/2019 09:22:25.171 [P2P] warn: could not open connection to peer (0243f9ff226cda8d5f00a293c56f4406fc9d7ff80d2184bd986c8513756c6e9b61): cannot attempt connection to self
xud_1        | 27/09/2019 09:22:25.172 [P2P] info: Peer 0243f9ff226cda8d5f00a293c56f4406fc9d7ff80d2184bd986c8513756c6e9b61 socket closed
xud_1        | 27/09/2019 09:22:25.175 [P2P] verbose: Verified reachability of advertised address: 0.0.0.0:8885
xud_1        | 27/09/2019 09:22:25.178 [RPC] info: gRPC server listening on 0.0.0.0:8886
```

**Comment by kilrau** [comment]
And this one is serious @sangaman : the seed was printed in xud log by seed util

Please check this log in details and make sure this can't happen in a separate PR: https://paste.ubuntu.com/p/JP4h5yTdqx/

**Comment by kilrau** [comment]
The above are to be fixed in separate PRs since not docker-related. Summary what needs to be fixed in this PR:
* Remove this line break for `Re-enter password:` 
image

* Fix stuck environment when exiting and re-entering xud ctl after successful wallet creation
image

**Comment by erkarl** [comment]
> xud logs show grpc server shutdown in between createNode and node key creation

This is expected behaviour since we shut down the `InitService` and start the regular gRPC `Service`.

> I managed to have a case where only the ltc wallet initialization succeeded. 

This should be solved by what @sangaman is working on - we'll wait for all the configured lnds to become responsive before attempting create.

> the seed was printed in xud log by seed util

https://github.com/ExchangeUnion/xud/issues/1254

**Comment by kilrau** [comment]
All 3 issues from https://github.com/ExchangeUnion/xud-docker/pull/134#issuecomment-535873630 are fixed. Tested mainnet and testnet.

Merging.

**Pull request merged**